### DISCUSSION TITLE
How to apply dynamic field defaults (for multiple fields)?

### DISCUSSION THREAD

**Discussion by andrenarchy** [directus/directus#10986] [Archive]
We have a collection with `author` data and the first step is to pick a person from a related `person` collection and then the author's data at the time of publication is entered:

![Screenshot from 2022-01-12 08-51-14](https://user-images.githubusercontent.com/1874116/149086029-71e13f8f-e72d-41f9-82f4-03c529ba86c0.png)

In 99% of the cases, the data is identical to what's stored on the `person`. Is there a way to add a button "apply defaults" that copies data from the `person` into author fields? 

It would need to know which fields go where so that'd be implemented in JS. The question is really how to get a button in there that can trigger code for modifying the state of multiple fields (based on the current state). Any ideas?

**Comment by oreilles** [comment]
That is not possible at the moment - the interfaces are not allowed to update multiple fields, so you won't be able to build it with an extension either.
I would suggest using a database trigger instead to auto fill the empty field at insert time - or maybe reconsider your design (do you really need thosee fields to be filled, if they're identical to the person's ones ?)

**Reply by andrenarchy** [reply]
Redesigning is not an option because the `person` data might change but the author name (on a publication) doesn't.

The trigger solution comes with its own drawbacks: it's not transparent to the user what's going on. They don't even see the applied values because they're only applied after saving.

**Reply by benhaynes** [reply]
I can see the benefit here (including for dynamic _slugs_), and would love to hear thoughts from the @directus/team on how this might be implemented in the no-code app (improves Cloud usability). Different options:

* Enable support within Interfaces, this requires code access
* Similar configuration to collection-level validation
* Field-level configuration
* Data flows

**Comment by benhaynes** [comment]
Would _Project Waterway_ work for this,  @rijkvanzanten ? It should on before-save, but would it ever be possible to add a client side trigger for staged field values in the app? 🤔

**Reply by andrenarchy** [reply]
Now I'm curious – what's Project Waterway? :sunglasses:

**Reply by rijkvanzanten** [reply]
@benhaynes I guess a custom action button trigger could be injected on the item detail page, which sends the currently _staged_ payload up to the intake webhook, which then gets transformed and returned? Needs a lot more thinking through and R&D, but that could certainly be an approach

**Reply by rijkvanzanten** [reply]
That could also be used for things like auto-translate on edit using an external service for example, so totally something we should investigate

**Reply by andrenarchy** [reply]
Would it be possible to provide an `action` button that can trigger an arbitrary function with the currently staged payload and which can set any field in the staged payload? If the handler is async it could also call external services for the use-case you suggested.

**Reply by rijkvanzanten** [reply]
Yeah! That's what I meant above 🙂 It's not currently a thing yet, but that would be incredibly powerful to have.

**Reply by andrenarchy** [reply]
Perfect! :) I got confused because you mentioned "webhook", that didn't seem necessary.

**Reply by andrenarchy** [reply]
Can this be implemented with an extension already? Or does it require a change in Directus' core?

**Reply by rijkvanzanten** [reply]
This would require a core change. There's currently no extension type that's able to inject custom header action buttons on the item detail pages like this.

That being said, you _might_ be able to make it work with a custom Interface that sends the current staged values to a custom endpoint, that then async runs the logic, responds with the updated staged values, which your custom interface then injects to all fields on the page. That being said, I've never tried that before, so your mileage may vary

**Reply by andrenarchy** [reply]
Regarding the button: it doesn't need to sit in the header, it can also be placed like a normal field.

For your custom interface suggestion: can a field interface inject values for other fields? How'd that work?

**Reply by rijkvanzanten** [reply]
All the values are stored in a reactive object called `values` that's provided by the parent form, so in your custom interface, you should be able to do `inject('values')` to pull that in. From there, you might be able to change that object to have the other interfaces reactively update. That being said, I've never tried it, so be aware it might not work 😁

**Reply by rgieseke** [reply]
So with the following one gets a button and can read and change values, e.g. put something in "surname" and "given name", copy it into "name". 

What I haven't figured out is how to send back the changed values to the parent. Is that possible without changing something in v-form? Not sure whether i need to trigger an event or am able to somehow call an update function of the parent to get the values changed in the actual form fields.

```ts
import InterfaceComponent from "./interface.vue";

export default {
  id: "custom-autofill",
  name: "Custom Autofill",
  icon: "box",
  description: "This is my custom autofill",
  component: InterfaceComponent,
  options: null,
  hideLabel: true,
  types: ["alias"],
  localTypes: ['presentation'],
  group: "presentation",
};
```

```vue
<template>
  <v-button large @click="applyDefaults()"> Apply defaults </v-button>
</template>

<script lang="ts">
import { ref, inject } from "vue";

export default {
  props: {},
  inject: ["values"],
  setup(props, { emit }) {
    const values = inject("values", ref<Record<string, any>>({}));
    return { applyDefaults };

    function applyDefaults(value) {
      console.log(values.value);
      values.value[
        "name"
      ] = `${values.value["given_name"]} ${values.value["surname"]}`;
      console.log(values.value);
    },
  },
};
</script>
```

**Reply by nickreese** [reply]
Did you figure this out? Trying to figure out how values would be passed to the parent now.

**Reply by nickreese** [reply]
So for anyone stumbling on this thread. 

```
emits: ["setFieldValue"],
```

Then Later:

```
emit("setFieldValue", { field: "address", value: "test updating" });
emit("setFieldValue", { field: "other", value: "test updating" });
```

**Reply by rgieseke** [reply]
> So for anyone stumbling on this thread. 
> 
> 
> 
> ```
> 
> emits: ["setFieldValue"],
> 
> ```
> 
> 
> 
> Then Later:
> 
> 
> 
> ```
> 
> emit("setFieldValue", { field: "address", value: "test updating" });
> 
> emit("setFieldValue", { field: "other", value: "test updating" });
> 
> ```

Does that work now? Great, I had the problem that it would only update with the last event sent.
Otherwise I was still using the hack above for multiple fields.

**Reply by nickreese** [reply]
This was working this morning (or was user error) it now only is updating the last value.

**Comment by andrenarchy** [comment]
If multi-field updating is not possible right now: is there a nice way to implement dynamic defaults for a single field? Would I have to copy the entire `input` interface code and create a new one with it or is there maybe a way to wrap it?

**Comment by luciandavidescu** [comment]
Great question, also had it in mind but didn't dare ask yet.

In publishing, there's always this almost full overlap between 'users' (as system items) and 'authors' (as public taxonomies).

Have seen my fair share of awful WP implementations where – on theme change – admins, webmasters or editors unwittingly became 'authors' of articles that they didn't actually write, only because they were the ones who uploaded the thing at some point. 

And this was  because some random developer at some random point chose not to mix internal user logic with external author display, so they made a random custom field for authorship. Then their thinking got lost along the way and disaster struck.

Sometimes it takes a while to discover, to great private frustration ("he literally stole my story!") or public embarassment ("sooo... content doesn't really match the name, who exactly is this writing the piece?''). 

So, hard one.... Users should BE and NOT be authors/collections at the same time. I guess the replication/"apply defaults" approach might make sense.

**Reply by oreilles** [reply]
In this kind of scenario, you would usually model your database with a table `user` holding all shared columns - and another table for each type of user, which will only hold the columns specifics to that kind of user. Then, depending on preferred use,  either:
- The base user table should have a foreign key to each of the user type columns
- Each of user-type tables should have a foreign key to the base user.
- In Directus, a many-to-any would be a perfect use case for this

**Reply by luciandavidescu** [reply]
Wow, dead simple... sorry, I'm only only now discovering lots of these things, it's overwhelming

**Reply by rijkvanzanten** [reply]
cc @erondpowell

**Comment by andrenarchy** [comment]
Just a quick update: @rgieseke and I found the following workaround:

`index.ts`:
```typescript
import { defineInterface } from '@directus/extensions-sdk'
import InterfaceComponent from './interface.vue'

export default defineInterface({
  id: 'apply-defaults-button',
  name: 'Apply Defaults Button',
  icon: 'box',
  description:
    'A button that applies defaults to fields based on other fields.',
  component: InterfaceComponent,
  options: null,
  hideLabel: true,
  types: ['alias'],
  localTypes: ['presentation'],
  group: 'presentation',
})
```

`interface.vue`:
```vue
<template>
  <v-button large @click="applyDefaults()">Apply defaults</v-button>
</template>

<script lang="ts">
import { ref, inject, getCurrentInstance } from "vue"

export default {
  props: {},
  inject: ["values"],
  setup() {
    const values = inject("values", ref<Record<string, any>>({}))
    const instance = getCurrentInstance()

    function applyDefaults() {
      const _person = values.value["person"]
      const person = !_person ?
        undefined :
        typeof _person === 'number' ?
        // TODO: fetch person from collection
        { given_name: 'TO', surname: 'DO', name: 'TO DO'} :
        _person

      const update = person && {
        ...values.value,
        given_name: person.given_name,
        surname: person.surname,
        name: person.name,
      }

      if (update)
        instance?.parent?.parent?.parent?.emit('update:modelValue', update)
    }

    return { applyDefaults }
  },
}
</script>
```

The hacky bit here is to call `instance?.parent?.parent?.parent?.emit('update:modelValue', update)`. This needs knowledge about how the component sits in the component tree. Currently, only the `input` event is handled on the parent `form-field-interface` which then bubbles up as `update:model-value` in `form-field` and `v-form` which ultimately emits `update:modelValue` with an object, see https://github.com/directus/directus/blob/29c450bfebfa69a167f1a4c3e7fcaabb1e9a0943/app/src/components/v-form/v-form.vue#L299-L303

All that we'd need for supporting this in a clean way is a new event that is handled in `form-field-interface` (and upwards). Maybe `update-fields`? What do you think about adding this?

**Reply by joselcvarela** [reply]
Right now, the populated fields comes from `edits`:
https://github.com/directus/directus/blob/e258f3d2889f612d7ece547b96d6c4c486805b02/app/src/composables/use-item/use-item.ts#L48

This store is hosted on app side. Maybe if make this store shared it will be easier to update multiple fields.
But, we may also want to add this to Cloud, so maybe the better approach is to create an interface into core to do this kind of job, because it also could allow to connect to Project Waterway.
Am I right, @rijkvanzanten ?

**Reply by rijkvanzanten** [reply]
Yeah absolutely 👍🏻

**Comment by Dupflo** [comment]
This would be a great adding. I am working on an app where the user is associated to a team and my backend is managing all the team inside different frontend apps (one for each team). So I can without any problem filter the available info based on a team TOKEN I create and set to the FE .env. 

The fact of saving by default the team associate to the user once a blog article is created for example instead of selecting manually with an M2O relationship would be a nice improvement for the User Experience

**Comment by lluishi93** [comment]
Hi there! I'm not able to update a translations type field with `setFieldValue`. I'm doing it this way:

```js
this.$emit('setFieldValue', { field: 'translations', value. [{ "id": 1, "foo_id": 1, "name": "change text", "languages_code": { "code": "es-es" } }] })
```

Is there something I'm doing wrong?
